va/caa-util.go

```diff
@@ -0,0 +1,190 @@
+	if len(encodedRDATA) < 2 {
+		// *very* badly formatted record, discard
+		return nil
+	}
+
+	// first octet is uint8 flags
+	flag := uint8(encodedRDATA[0])
+
+	// second octet is uint8 length of tag
+	tagLen := uint8(encodedRDATA[1])
+
+	if uint8(len(encodedRDATA)) < 2+tagLen {
+		// badly formatted record, discard
+		return nil
+	}
+	tag := string(encodedRDATA[2 : 2+tagLen])
+
+	// Only decode tags we understand, value/valuebuf can be empty
+	// (that would be weird though...)
+	var valueBuf []byte
+	var value string
+	if tag == "issue" || tag == "issuewild" || tag == "iodef" {
+		value = string(encodedRDATA[2+tagLen:])
+	} else {
+		// unknown tag value may not be string, so store in buf
+		valueBuf = encodedRDATA[2+tagLen:]
+	}
+
+	return &CAA{flag: flag, tag: tag, valueBuf: valueBuf, value: value}
+}
+
+// contains returned CAA records filtered by tag.
+type CAASet struct {
+	issue     []*CAA
+	issuewild []*CAA
+	iodef     []*CAA
+	unknown   []*CAA
+}
+
+// returns true if any CAA records have unknown tag properties and are flagged critical.
+func (caaSet CAASet) criticalUnknown() bool {
+	if len(caaSet.unknown) > 0 {
+		for _, caaRecord := range caaSet.unknown {
+			// Critical flag is 1, but acording to RFC 6844 any flag other than
+			// 0 should currently be interpreted as critical.
+			if caaRecord.flag > 0 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func newCAASet(CAAs []*CAA) *CAASet {
+	var issueSet []*CAA
+	var issuewildSet []*CAA
+	var iodefSet []*CAA
+	var unknownSet []*CAA
+
+	for _, caaRecord := range CAAs {
+		switch caaRecord.tag {
+		case "issue":
+			issueSet = append(issueSet, caaRecord)
+		case "issuewild":
+			issuewildSet = append(issuewildSet, caaRecord)
+		case "iodef":
+			iodefSet = append(iodefSet, caaRecord)
+		default:
+			unknownSet = append(unknownSet, caaRecord)
+		}
+	}
+
+	return &CAASet{issue: issueSet, issuewild: issuewildSet, iodef: iodefSet, unknown: unknownSet}
+}
+
+func lookupCNAME(client *dns.Client, server, domain string) (string, error) {
+	m := new(dns.Msg)
+	m.SetQuestion(domain, dns.TypeCNAME)
+	m.SetEdns0(4096, true)
+	r, _, err := client.Exchange(m, server)
+	if err != nil {
+		return "", err
+	}
+
+	for _, answer := range r.Answer {
+		if cname, ok := answer.(*dns.CNAME); ok {
+			return cname.Target, nil
+		}
+	}
+
+	return "", nil
+}
+
+func getCaa(client *dns.Client, server string, domain string, alias bool) ([]*CAA, error) {
+	if alias {
+		canonName, err := lookupCNAME(client, server, dns.Fqdn(domain))
+		if err != nil {
+			return nil, err
+		}
+		if canonName == "" || canonName == domain {
+			return []*CAA{}, nil
+		}
+		domain = canonName
+	}
+
+	m := new(dns.Msg)
+	m.SetQuestion(dns.Fqdn(domain), dns.TypeCAA)
+	m.SetEdns0(4096, true)
+	r, _, err := client.Exchange(m, server)
+	if err != nil {
+		return nil, err
+	}
+
+	var CAAs []*CAA
+	for _, answer := range r.Answer {
+		if answer.Header().Rrtype == dns.TypeCAA {
+			caaLenStr := strings.Fields(answer.String())[5]
```

Comment:
You should probably check the length of the Fields result before indexing into it, just to be paranoid.